### ISSUE TITLE
[AutoSuggest for "Service type"] Step 3 - Build the FE AutoSuggest interface for FL

### ISSUE THREAD

**Issue by davidmpickett** [department-of-veterans-affairs/va.gov-cms#18911]
**[2024-08-16]** Need answers to open questions in ticket body and to get those into requirements / ACs

## Description or Additional Context 
In #18909 and #18910 we built a method for filtering. Now we will integrate that mechanism into the FE.

[Finalized autosuggest Figma designs](https://www.figma.com/design/GvNVeG9GAXZaBpny7vNxZP/Facility-Locator?node-id=1496-38390&t=DMVdnIDuv3oQB4FM-1)

## Steps for Implementation
Questions that were previously in this section are answered in the UX ticket to finalize designs: #18960 

## Acceptance Criteria
- [x] Works with conditional rendering from progressive disclosure
- [x] Only needs to be for VA Health Facilities
- [x] Minimum characters before matches are filtered: 2
- [x] Between 0-1 characters, all VAMC services show
- [x] If a matching service is not found while the field is active / in-use, error message of "No results found" should display in the dropdown
- [x] All suggestions that match should display in the dropdown, no limit
- [x] Continue to winnow suggestions when user enters additional letters after suggestions are displayed
- [x] Display full service name plus AKA (`Name (Known as)`) where the AKA exists; no character limit
- [x] Show "All VA health services" in the dropdown as the top item before the user types, and always show it at the top of the dropdown by default
- [x] Service type input should not be a required input
- [x] If the user has interacted with the Service type input but has not selected a valid service and clicks Search, run the search with "All VA health services."
- [x] Before any input is typed, all unfiltered services should appear in the dropdown with "All VA health services" at the top
- [x] If a user types a search term and then clears it, all unfiltered services should appear in the dropdown again with "All VA health services" at the top
- [x] Uses the combo box as noted in #18960 
- [x] Requires design review
- [x] Requires accessibility review
- [x] Feature flag for deployment to STAGING ONLY on branch with progressive disclosure for facility locator useability

**Comment by jilladams** [comment]
@Agile6MSkinner noting: Eli's work leading up to this is now shipped, so we can pull this into sprint as soon as we can get it refined on the product side.

**Comment by Agile6MSkinner** [comment]
Test plan and QA ticket for combined FL Q1-25 iteration epic needs to be created.

**Comment by randimays** [comment]
@thejordanwood @davidmpickett @eselkin I have many questions:

1. Field is active, dropdown is open ([design](https://www.figma.com/design/GvNVeG9GAXZaBpny7vNxZP/Facility-Locator?node-id=45-15352&t=JTV1Ycs8BzThUnX2-0))
  - How are **Popular Services** determined?
  - **Popular Services** text should not be clickable, correct? I'm not sure this is possible given the combo box code ([USWDS](https://designsystem.digital.gov/components/combo-box/) / [VADS](https://design.va.gov/storybook/?path=/docs/uswds-va-combo-box--docs))
2. Field is active, dropdown is open, search term is entered ([design](https://www.figma.com/design/GvNVeG9GAXZaBpny7vNxZP/Facility-Locator?node-id=45-16987&t=JTV1Ycs8BzThUnX2-0)). Confirming the "See results for all VA health services" option should always show in the dropdown, even though it doesn't in the design comp I linked?
3. Technical question: how do I get the `va-healthcare-services.json` to work locally? I tried building content-build and I'm getting a 404 on that file. It's always 404'd for vamc-ehr.json as well.
4. Technical question for matching logic: Looking at [the Defined Behaviors](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/18960) > `Text displayed when matching on data other than Term Name`: Are the below examples of search terms to expect dropdown option(s) correct? Referencing [va-healthcare-services.json](https://www.va.gov/data/cms/va-healthcare-services.json)...
  
  - **Term**: "heart"
  - **Options**:
    - "Cardiology (Heart and circulation)"
    - "Transplant surgery"
<br />

  - **Term**: "heart disease"
  - **Option**:"Cardiology (Heart Disease)"
<br />

  - **Term**: "exercise"
  - **Options**:"Recreation and creative arts therapy"
<br />

  - **Term**: "agent orange"
  - **Options**:"Toxic exposure screening (Military exposure screening, war-related exposure screening)"?
<br />

  - **Term**: "cardiology"
  - **Options**: "Cardiology"
<br />

  - **Term**: "cancer"
  - **Options**:?? This one's tricky because we get a lot of matches on the associated conditions such as skin cancer, prostate cancer, etc.

5. The combo box in the design system is coded to say "No results found" when it isn't passed any options. Per our **Defined Behaviors**, we don't want to start filtering services until there are at least 3 characters. Is the interaction in the below video OK? Note: I realize it's not updating once there are 3 characters; I'm still working on getting that part to function properly.

https://github.com/user-attachments/assets/ae9150cc-7879-4f27-9471-9cd2a9ff6a76

**Comment by thejordanwood** [comment]
@randimays Apologies for the confusion, this ticket had outdated information and was pointing to an older version of the designs. I've updated that in the ticket. 

- [Finalized Figma designs](https://www.figma.com/design/GvNVeG9GAXZaBpny7vNxZP/Facility-Locator?node-id=1496-38390&t=zrSwX3l0EuNOPH2r-1)
- [Ticket with defined behaviors for autosuggest that are hard to show in designs](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/18960)

1. Popular services are no longer MVP. These can be ignored for now. 
2. The newer designs show how "All VA health services" should look as the default option. It should show as the default when first loading the Service Type field ([example](https://www.figma.com/design/GvNVeG9GAXZaBpny7vNxZP/Facility-Locator?node-id=1613-14475&t=mg9Avxyqgx7TkqwC-1)), then it should always display in the dropdown ([example](https://www.figma.com/design/GvNVeG9GAXZaBpny7vNxZP/Facility-Locator?node-id=1824-6907&t=mg9Avxyqgx7TkqwC-1)). Hopefully that clears up this question, but if not just let me know! 
3. No comment
4. No comment
5. In [this ticket](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/18960), we decided that there should be at least 2 characters in the field before matches are displayed. Is there a way to not show the "No results found" until 2 characters are entered? If not, we may need to rethink this condition and go with a 1 character minimum before matches are displayed. cc: @davidmpickett

**Comment by randimays** [comment]
@Agile6MSkinner [Number 4](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/18911#issuecomment-2603269112) in my list of questions will need a Product review, please.

**Comment by davidmpickett** [comment]
cc @Agile6MSkinner 

5) I think we should go to usability testing with it showing matches as soon as 1 character matches.

- Displaying "no results found" is not honest if there are results that we are suppressing
- Since that's the default behavior of the Combo Box and we are the first VA team to implement it, it would be good for us to test the behavior as is and share those findings with DST.

**Comment by davidmpickett** [comment]
> 4. Technical question for matching logic: Looking at [the Defined Behaviors](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/18960) > `Text displayed when matching on data other than Term Name`: Are the below examples of search terms to expect dropdown option(s) correct? Referencing [va-healthcare-services.json](https://www.va.gov/data/cms/va-healthcare-services.json)...
> 
> * **Term**: "heart"
> * **Options**:
>   
>   * "Cardiology (Heart and circulation)"
>   
>   * "Transplant surgery"

Correct, and should also show: 
* "Cardiovascular surgery (Heart surgery)"


> * **Term**: "heart disease"
> * **Option**:"Cardiology (Heart Disease)"

Correct, but would still display as:
* "Cardiology (Heart and circulation)"

> * **Term**: "exercise"
> * **Options**:"Recreation and creative arts therapy"
> 
> 
> * **Term**: "agent orange"
> * **Options**:"Toxic exposure screening (Military exposure screening, war-related exposure screening)"?
> 
> 

Correct and correct

> * **Term**: "cardiology"
> * **Options**: "Cardiology"
> 

Correct, but would still display as:
* "Cardiology (Heart and circulation)"

> * **Term**: "cancer"
> * **Options**:?? This one's tricky because we get a lot of matches on the associated conditions such as skin cancer, prostate cancer, etc.
> 

This is a great example because it covers the whole hierarchy of possible fields to match on
1. Results that match on Term Name [e.g. Cancer care]
2. Results that match on Patient Friendly Name [e.g. "Surgical oncology (Cancer surgery)"]
4. Results that match on Common Conditions [e.g. "Dermatology (Skin conditions and diseases)"]
5. Results that match on Description [e.g. "Hematology/oncology (Blood disorders and medical oncology")

**Comment by randimays** [comment]
@davidmpickett Thank you! I think that cleared up my questions for matching behavior.

**Comment by randimays** [comment]
@thejordanwood We aren't able to customize the "No results found" messaging with the combo box.

<img width="521" alt="Image" src="https://github.com/user-attachments/assets/dcf87e66-40f3-4e15-95e9-12dda93c3531" />

I can put in a DST ticket to allow customization for this and we can tackle at that point, if you like?

**Comment by thejordanwood** [comment]
@randimays Let's go with [the comment that Dave made](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/18911#issuecomment-2605254014). We can show matches as soon as 1 character matches and usability test this. No need to put in a DST ticket or add customization!

> 1. I think we should go to usability testing with it showing matches as soon as 1 character matches.
> - Displaying "no results found" is not honest if there are results that we are suppressing
> - Since that's the default behavior of the Combo Box and we are the first VA team to implement it, it would be good for us to test the behavior as is and share those findings with DST.

**Comment by randimays** [comment]
@thejordanwood I don't have the ability to keep that `No results found` messaging from popping up if the search term doesn't match any services in our list. I'm not sure I follow this:

> Displaying "no results found" is not honest if there are results that we are suppressing

**Comment by thejordanwood** [comment]
@randimays 

- We should only show the "no results found" messaging if there are no matches. It's okay if it pops up after typing in 1 or more characters, as long as there are truly no matches for any services in our list. 
- I believe what Dave was saying with that comment is in your video, you typed in "m" then typed in "me", which showed the "no results found" messaging. In that case, this messaging is not true because there are services in our list that match those results, but they were being suppressed when we tried to add a minimum character count. This is why we're now suggesting to go with the default behavior of the Combo Box, which is to have the minimum character count as 1 instead of the original 2 we had suggested.

**Comment by jilladams** [comment]
Need to remove API ID from matching logic when we come back to this:
https://github.com/department-of-veterans-affairs/vets-website/blob/7977461a7eb3148a98ec596f7d9c7d84a5916ef6/src/applications/facility-locator/hooks/useServiceType.jsx#L37

**Comment by jilladams** [comment]
I've documented the business logic about field matching / weighting here: https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/products/facilities/facility-locator/engineering#services-autosuggest.

**Comment by randimays** [comment]
This ticket is still blocked; I have not worked on it at all this sprint (23).

**Comment by randimays** [comment]
Eli just merged the autosuggest work this morning so I'm picking this back up now.

**Comment by randimays** [comment]
@davidmpickett Could you clarify these two AC please? Should we be showing `Name (Known as)` for **all** matches, or only the ones that do not match on the name?

<img width="772" alt="Image" src="https://github.com/user-attachments/assets/1dc0ed65-c5c3-4ae0-bea7-19aaf9151972" />

**Comment by davidmpickett** [comment]
Always show `Name (Known as)` for all items. However, not every item will have `(Known as)` From the Figma, this is true even before the user has typed a single character:

![Image](https://github.com/user-attachments/assets/c6b43631-6840-4fef-93a0-33330c00f9a3)

And after characters have been entered:

![Image](https://github.com/user-attachments/assets/653b562f-a117-4624-a595-433844e405e5)

[This list in Drupal](https://prod.cms.va.gov/va-services) is a great way to see which Services currently have a Patient Friendly Name / 'Known As'

**Comment by davidmpickett** [comment]
> Additional question: if the user:
> 
> 1. types something
> 2. doesn't select one of the matches
> 3. un-focuses the input
> 
> Should we retain what they typed and show an error (since we know that the search won't succeed)?

Good catch, it looks like that was not something Jordan covered in the [Error states](https://www.figma.com/design/GvNVeG9GAXZaBpny7vNxZP/Facility-Locator?node-id=172-16230). I think retaining what they've typed and showing an error (without waiting for them to click search) is smart. the error message could say "Select a service type" vs. the "Start typing and select a service type" message that shows if they click Search without typing anything.

**Comment by randimays** [comment]
Got it - thanks!

**Comment by randimays** [comment]
@davidmpickett I just noticed the Show results in alphabetical order above [this page in the Figma](https://www.figma.com/design/GvNVeG9GAXZaBpny7vNxZP/Facility-Locator?node-id=1613-15101&t=xrdZrq6aOSdUNY1A-0) - should that be the case all the time? I'm assuming we want the initial full list in alphabetical order, and then use the priority sorting that Eli built when there is a search term / matches.

**Comment by davidmpickett** [comment]
> [@davidmpickett](https://github.com/davidmpickett) I just noticed the `Show results in alphabetical order` above [this page in the Figma](https://www.figma.com/design/GvNVeG9GAXZaBpny7vNxZP/Facility-Locator?node-id=1613-15101&t=xrdZrq6aOSdUNY1A-0) - should that be the case all the time? I'm assuming we want the initial full list in alphabetical order, and then use the priority sorting that Eli built when there is a search term / matches.

Absolutely. The Figma file did not fully document the priority sorting. If there are terms with equal priority ranking, then those should be alpha sorted.

**Comment by randimays** [comment]
Update: very close to having this ready for PR review, just have Cypress tests left to add. The current state of e2e tests in the app is a little chaotic. It's difficult to tell where to add them. I'm trying to do a bit of organizing / cleanup to avoid stacking another test on top of everything - maybe I can work the tests into the existing tests. Hoping to have this ready to review by Monday morning.

**Comment by jilladams** [comment]
Amazing! For context: we are getting a pileup of testing. We will want to let @Shamiaalam finish the re-verification of Progressive Disclosure behind a flipper (on Staging) before we toss this her way. Not sure yet if she'll be ready for it by Monday or not. 

However: this is going to go to research anyway. So if code changes get cleared before we have all our ducks in a row for testing, I _think_ it will be ok to go ahead and merge behind a flipper. We will just need to make sure the flipper stays disabled through Prog Disclosure/ Mobile map staging review window (March 12-18), and we can test before / after.

@Agile6MSkinner fyi.

**Comment by Shamiaalam** [comment]
@jilladams Got it! I’ll focus on finishing the re-verification of Progressive Disclosure behind the flipper on Staging first. I’ll keep you posted on my progress if I’m ready to pick up the next round of testing by Monday morning,  I’ll let you know.

**Comment by randimays** [comment]
Update here: I created a PR that peeled off some of the changes for autosuggest to make that feature PR a bit smaller. That was merged a couple hours ago, so now I am getting the autosuggest feature PR ready for review. I had a couple of Cypress failures, so I fixed those. I needed to get the review instance running and go through the test plan one more time before passing it off for functional testing. I'll likely post this for review tomorrow.

**Comment by randimays** [comment]
Moving this back to In Progress while I troubleshoot an issue with the health services data and the Cypress tests in the pipeline. Hoping to have this resolved by EOD and get it posted for review, but troubleshooting things in the pipeline in particular is tricky and it's going to take time (for it to run).

**Comment by randimays** [comment]
Eli helped resolved the remaining Cypress issue and the PR has been posted for review 🥳

**Comment by davidmpickett** [comment]
> Between 0-1 characters, all VAMC services show

Documenting a current quirky behavior which I'm okay with leaving as is until after usability testing, but would want to make sure we try and get users to react to it if possible (cc @biannuz)

1) Type a single character "a" = see all results
2) Type two characters "ar" = see reduced results matching 2 characters
3) Delete a character = results remain reduced to match 2 characters rather than reverting to all results

![Image](https://github.com/user-attachments/assets/ccaeb30a-f299-41d4-bf04-827c8bdb234b)

![Image](https://github.com/user-attachments/assets/fea7361e-11bf-4518-a29e-81d768d02813)

![Image](https://github.com/user-attachments/assets/46e0b682-1a13-4bfa-9223-67ab4235fdd3)

**Comment by randimays** [comment]
@davidmpickett Whoops - I had addressed this after we noticed it in Demo, but that change got lost while I was troubleshooting Cypress stuff for the past few days. I can get it fixed in the current branch.

**Comment by davidmpickett** [comment]
Hey @randimays, just compared this side-by-side with [Figma design](https://www.figma.com/design/GvNVeG9GAXZaBpny7vNxZP/Facility-Locator?node-id=1824-7494&t=8Eius0shul8sfyoA-4) and there are a few things I'm noticing. Some of these may be out of scope, but wanted to call them out so we make sure they get addressed/ticketed cc @Agile6MSkinner @jilladams 

1. The "T" in "Facility Type" label should be lowercase
2. The [error message when no Facility type is selected](https://www.figma.com/design/GvNVeG9GAXZaBpny7vNxZP/Facility-Locator?node-id=172-20935&t=8Eius0shul8sfyoA-4) should read "Select a facility type"
3. Service type field should have hint text "Begin typing to search for a service, like like vision or dental"
4. Text below the search box should read "Enter a location (street, city, state, or zip code) and facility type, then search to find facilities."
5. When there are no results found on mobile, the List view and Map view should have [different text for suggestions ](https://www.figma.com/design/GvNVeG9GAXZaBpny7vNxZP/Facility-Locator?node-id=752-14458&t=8Eius0shul8sfyoA-4)

**Comment by davidmpickett** [comment]
@Agile6MSkinner @jilladams This one is definitely out of scope for @randimays ticket, but doesn't seem to be ticketed elsewhere

6. If a user clicks "Use my location" but doesn't have location sharing enables, the error message shown should match the text [in Figma](https://www.figma.com/design/GvNVeG9GAXZaBpny7vNxZP/Facility-Locator?node-id=172-21049&t=8Eius0shul8sfyoA-4)

**Comment by davidmpickett** [comment]
7. If a user selects "Community providers" and attempts to search without selecting a service type, the error message should [match Figma](https://www.figma.com/design/GvNVeG9GAXZaBpny7vNxZP/Facility-Locator?node-id=178-32761&t=8Eius0shul8sfyoA-4)

**Comment by davidmpickett** [comment]
8. The Design System just merged work that will allow the Search Button to be Full Width! Oh wait @jilladams just made a [separate ticket](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/20837) for that.

**Comment by eselkin** [comment]
Yeah. These should be other tickets. Well the location permission and the CCP stuff should be.

**Comment by randimays** [comment]
The code for this feature has been successfully deployed to all environments. Note that the flag is **ON** in staging, so the feature can be viewed/tested there (cc @Shamiaalam @jilladams). The flag is **OFF** in production, so services autosuggest cannot be viewed there.

Closing this as complete, noting that all but [#8](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/18911#issuecomment-2711177202) of Dave's above comments will be addressed in a fast-follow PR [here](https://github.com/department-of-veterans-affairs/vets-website/pull/35141).
